Pluggable InAppBrowser on Android

dvicory opened the pull request.

dvicory committed:
Mostly compatible version of InAppBrowser for Crosswalk.
- The new strategy for this is to create a new CordovaWebViewImpl instead of relying
  directly on WebView. This abstracts the web view and removes the need to do some things
  that Cordova already does.
  - The new Cordova webview inherits the preferences already set in config.xml. It will use
    the same engine already specified by the webView preference. Plugins are cleared so any
    plugins you define in config.xml will not be available in the InAppBrowser.
  - To get the necessary events (loadstart, loadstop, loaderror, exit) and handle existing
    InAppBrowser functionality (such as opening the dialer when click on tel links), the
    InAppBrowserInternal plugin was created which is only loaded within the InAppBrowser Cordova webview.
  - For security reasons, the Cordova native bridge is disabled for sites inside InAppBrowser.
- There is not an API accessible getSettings for Crosswalk, thus zoom controls
  and database storage toggle are not supported when using Crosswalk. This would require
  upstream changes to Crosswalk to make those settings accessible.
  See https://crosswalk-project.org/jira/browse/XWALK-3313.
- Callbacks are only supported for Crosswalk as it requires use of the @JavascriptInterface
  annotation. There is android.webkit.JavascriptInterface and org.xwalk.core.JavascriptInterface
  though it does not seem possible to use both. Being able to support either at runtime would
  require Java byte code twiddling, selecting the engine at build time, or using another method
  of JS->Native communication altogether (for instance, an iframe with a changing src).

dvicory committed:
Beginnings of making a pluggable plugin.

dvicory committed:
Remove setDriver on InAppBrowserInternal and pass the driver to init

dvicory committed:
Move the creation of the CordovaWebViewImpl and applying the settings to their own methods

dvicory committed:
Use XML resource IDs to quell warnings; remove unneeded edittext listener; fix issue with vertical scrolling in edittext

uareurapid committed:
handle app store urls in system browser

jonkessler commented on src/android/InAppBrowserDriver.java:
```diff
@@ -0,0 +1,543 @@
+        PluginManager pluginManager = null;
+        try {
+            pluginManager = (PluginManager)InAppBrowserImpl.invokeOn(inAppBrowser.getWebView().getView(), "getPluginManager", new Class[0], new Object[0]);
+        } catch (InAppBrowserImpl.InvokeException e) {
```
Could we put a log or something in this empty catch?

dvicory replied:
:+1:

jonkessler commented on src/android/InAppBrowserDriver.java:
```diff
@@ -0,0 +1,543 @@
+import android.widget.LinearLayout;
+import android.widget.RelativeLayout;
+
+import com.appfolio.appfolio_property_manager.R;
```
Is this import correct?

dvicory replied:
No, this would have been caught if I created a fresh Cordova project to develop with. And run InAppBrowser's test suite (still need to figure that out). :+1:

jonkessler commented on src/android/InAppBrowserDriver.java:
```diff
@@ -0,0 +1,543 @@
+    public void injectDeferredObject(String source, String jsWrapper) {
+        String scriptToInject;
+        if (jsWrapper != null) {
+            org.json.JSONArray jsonEsc = new org.json.JSONArray();
```
We shouldn't need to fully specify `org.json.JSONArray`, right?

dvicory replied:
:+1: Copy/paste didn't cause Android Studio to remove the fully qualified class name.

jonkessler commented on src/android/InAppBrowser.java:
```diff
@@ -1,724 +1,45 @@
+    String HARDWARE_BACK_BUTTON = "hardwareback";
+
+    InAppBrowser getInAppBrowser();
+    CordovaInterface getCordova();
```
Perhaps a better name would be `getCordovaInterface`?

dvicory replied:
:+1:

dvicory committed:
Add log if getPluginManager fails and simplify return logic

dvicory committed:
Don't use fully qualified class name for JSONArray

dvicory committed:
Rename getCordova to getCordovaInterface

jcesarmobile committed:
CB-10305 Gray bar appears in the wrong place on iOS

This closes #137

birge committed:
CB-7679 add fix for iOS upload. This closes #139

jcesarmobile committed:
Fixes CB-10607

When the device is in portrait upside down the toolbar appears at the
bottom. This fix set the origin to the 0,0 point
